### PR TITLE
[feat](Nereids): Add hint `NTH_OPTIMIZED_PLAN` to let the optimzier can select n-th optimized plan

### PR DESCRIPTION
**Pull request by keanji-x**
# Proposed changes

Add hint NTH_OPTIMIZED_PLAN to let the optimzier can select n-th optimized plan. For example:
you could use 
```
select /*+SET_VAR("nth_optimized_plan"=2) */ * from table
```
to select the second-best plan in the optimizer.

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 1]
```diff
@@ -706,4 +710,172 @@ public String toString() {
...
+        if (groupExpression.getPlan() instanceof LeafPlan) {
+            res.add(Pair.of(0L, groupExpression.getCostByProperties(prop)));
+            return res;
+        }
+
+        double bestChildrenCost = 0;
+        List<List<Pair<Long, Double>>> children = new ArrayList<>();
+        for (int i = 0; i < inputProperties.size(); i++) {
+            Preconditions.checkArgument(!groupExpression.child(i).equals(groupExpression.getOwnerGroup())
+                    || !prop.equals(inputProperties.get(i)));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:769]: the check for enforce props is hard to understand, add some comment to explain it~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 2]
```diff
@@ -706,4 +710,172 @@ public String toString() {
...
+            }
+            res.add(Pair.of(c.first,
+                    childCost + groupExpression.getCostByProperties(prop) - bestChildrenCost));
+        }
+        return res;
+    }
+
+    // we permute all children, e.g.,
+    //      for children [1, 2] [1, 2, 3]
+    //      we can get: 0: [1,1] 1:[1, 2] 2:[1, 3] 3:[2, 1] 4:[2, 2] 5:[2, 3]
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:790]: use javadoc style~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 3]
```diff
@@ -706,4 +710,172 @@ public String toString() {
...
+            return;
+        }
+        for (int i = 0; i < children.get(index).size(); i++) {
+            List<Integer> next = new ArrayList<>(current);
+            next.add(i);
+            permute(children, index + 1, result, next);
+        }
+    }
+
+    private static long getUniqueId(List<List<Pair<Long, Double>>> lists, List<Integer> current) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:804]: add some comment to explain the logic of id generator

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 4]
```diff
@@ -250,6 +252,16 @@ public Group getRoot() {
         return cascadesContext.getMemo().getRoot();
     }
 
+    private PhysicalPlan chooseNthPlan(Group rootGroup, PhysicalProperties physicalProperties, int nthPlan) {
+        if (nthPlan == 1) {
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java (hunk 1 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:770]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java (hunk 2 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java (hunk 3 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:813]: done

**Review by morrySnow — APPROVED** [review]

**Pull request merged**